src/Aspire.Dashboard/Components/Pages/ConsoleLogs.razor.cs

```diff
@@ -126,57 +147,32 @@ private async ValueTask LoadLogsAsync()
         }
         else
         {
-            _watchLogsTokenSource = new CancellationTokenSource();
-            if (await _selectedResource.LogSource.StartAsync(_watchLogsTokenSource.Token))
+            var cancellationToken = await _logSubscriptionCancellationSeries.NextAsync();
+
+            var subscription = ResourceService.SubscribeConsoleLogs(_selectedResource.Name, cancellationToken);
+
+            if (subscription is not null)
             {
-                var outputTask = Task.Run(async () =>
-                {
-                    await _logViewer.WatchLogsAsync(
-                        () => _selectedResource.LogSource.WatchOutputLogAsync(_watchLogsTokenSource.Token),
-                        new LogParserOptions()
-                        {
-                            ConvertTimestampsFromUtc = ConvertTimestampsFromUtc
-                        }
-                    );
-                });
-
-                var errorTask = Task.Run(async () =>
-                {
-                    await _logViewer.WatchLogsAsync(
-                        () => _selectedResource.LogSource.WatchErrorLogAsync(_watchLogsTokenSource.Token),
-                        new LogParserOptions()
-                        {
-                            ConvertTimestampsFromUtc = ConvertTimestampsFromUtc,
-                            LogEntryType = LogEntryType.Error
-                        }
-                    );
-                });
-
-                _ = Task.WhenAll(outputTask, errorTask).ContinueWith((task) =>
-                {
-                    // If the task was canceled, that means one or both of the underlying tasks were canceled
-                    // which only really happens when we switch to another container source or when leaving
-                    // page. In both of those situations we can skip updating the status because it'll just
-                    // cause a flash of text change before it changes again or the page is navigated away.
-                    if (!task.IsCanceled)
-                    {
-                        _status = Loc[Dashboard.Resources.ConsoleLogs.ConsoleLogsFinishedWatchingLogs];
-                    }
-                }, TaskScheduler.Current);
+                var task = _logViewer.SetLogSourceAsync(
+                    subscription,
+                    convertTimestampsFromUtc: _selectedResource is ContainerViewModel);
 
+                _initialisedSuccessfully = true;
                 _status = Loc[Dashboard.Resources.ConsoleLogs.ConsoleLogsWatchingLogs];
+
+                // Indicate when logs finish (other than by cancellation).
+                _ = task.ContinueWith(
```

Comment:
Is there a reason to use ContinueWith vs async/await in an unawaited method? I guess there is the benefit or automatically avoid cancelltion exceptions.

Reply:
This form allows `Options.NotOnCanceled` as you say, which avoids overhead in the cancelled case. It also avoids having to schedule code that runs before the yield, so there's less stuff on the thread pool.